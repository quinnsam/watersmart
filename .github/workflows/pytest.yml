# see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

name: Pytest

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:
  schedule:
    - cron: "0 0 * * *"

jobs:
  build:

    runs-on: ubuntu-latest
    strategy:
      fail-fast: false
      matrix:
        python-version: ["3.12"]

    steps:
    - uses: actions/checkout@v4
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v5
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.test.txt
    - name: Test with pytest
      run: |
        pytest
    - name: Validate coverage
      run: |
        coverage json -q -o - | python3 .github/validate_coverage.py