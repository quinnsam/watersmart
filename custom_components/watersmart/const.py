"""Constants for the WaterSmart integration."""

from datetime import timedelta
from enum import StrEnum, auto
from typing import Final

ATTRIBUTION: Final = "Data scraped from WaterSmart"
DOMAIN: Final = "watersmart"
MANUFACTURER: Final = "WaterSmart by VertexOne"
DEFAULT_SCAN_INTERVAL = timedelta(hours=1)


class SensorKey(StrEnum):
    """Converter key enumeration class."""

    GALLONS_FOR_MOST_RECENT_HOUR = auto()
    GALLONS_FOR_MOST_RECENT_FULL_DAY_KEY = auto()
