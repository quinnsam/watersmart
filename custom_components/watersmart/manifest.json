{
  "codeowners": ["@wbyoung"],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/wbyoung/watersmart/",
  "domain": "watersmart",
  "iot_class": "cloud_polling",
  "name": "WaterSmart",
  "requirements": [],
  "version": "0.1.0"
}
