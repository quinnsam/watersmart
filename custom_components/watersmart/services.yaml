get_water_usage_hourly_history:
  fields:
    config_entry:
      required: true
      selector:
        config_entry:
          integration: watersmart
    cached:
      required: false
      default: false
      selector:
        boolean:
    start:
      required: false
      example: "2024-01-01 00:00:00"
      selector:
        datetime:
    end:
      required: false
      example: "2024-01-01 00:00:00"
      selector:
        datetime:
