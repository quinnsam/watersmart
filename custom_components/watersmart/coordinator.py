"""The WaterSmart coordinator."""

from asyncio import timeout
from collections.abc import Callable
import datetime as dt
import functools
import logging
from typing import Any, Protocol, TypedDict, cast

from aiohttp.client_exceptions import ClientConnectorError
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
from homeassistant.util.dt import as_local, get_default_time_zone, start_of_local_day

from .client import AuthenticationError, UsageRecord, WaterSmartClient
from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, MANUFACTURER, SensorKey
from .types import SensorData

EXCEPTIONS = (AuthenticationError, ClientConnectorError)

_LOGGER = logging.getLogger(__name__)


class CoordinatorData(TypedDict, total=False):
    """Shape of coordinator data."""

    gallons_for_most_recent_hour: SensorData
    gallons_for_most_recent_full_day: SensorData
    gallons_today: SensorData
    hourly: list[UsageRecord]


class _DataConverterT(Protocol):
    converter_key: SensorKey

    def __call__(self, data: CoordinatorData) -> SensorData: ...  # pragma no cover


class WaterSmartUpdateCoordinator(DataUpdateCoordinator[CoordinatorData]):
    """Class to manage fetching Watersmart data."""

    data_converters: tuple[_DataConverterT, ...] = ()

    def __init__(
        self,
        hass: HomeAssistant,
        watersmart: WaterSmartClient,
        hostname: str,
        username: str,
    ) -> None:
        """Initialize."""

        super().__init__(
            hass,
            _LOGGER,
            name=f"WaterSmart {hostname}",
            update_interval=DEFAULT_SCAN_INTERVAL,
        )

        self.watersmart = watersmart
        self.hostname = hostname
        self.username = username
        self.device_info = _get_device_info(hostname, username)
        self.data: CoordinatorData = {}
        self.data_converters = (
            _sensor_data_for_most_recent_hour,
            _sensor_data_for_most_recent_full_day,
            _sensor_data_for_today,
        )

    async def _async_update_data(self) -> CoordinatorData:
        """Update data via library.

        Returns:
            The updated data.

        Raises:
            UpdateFailed: If there is an error that could typically occur.
        """
        try:
            async with timeout(30):
                result: CoordinatorData = {
                    "hourly": await self.watersmart.async_get_hourly_data(),
                }
        except EXCEPTIONS as error:
            raise UpdateFailed(error) from error

        for converter in self.data_converters:
            cast("dict[str, SensorData]", result)[converter.converter_key] = converter(
                result
            )

        _LOGGER.debug("Async update complete")

        return result


def _get_device_info(hostname: str, username: str) -> DeviceInfo:
    """Get device info.

    Returns:
        The device info.
    """
    return DeviceInfo(
        entry_type=DeviceEntryType.SERVICE,
        identifiers={(DOMAIN, f"{hostname}-{username}")},
        manufacturer=MANUFACTURER,
        name=f"WaterSmart ({hostname})",
    )


def _from_timestamp(timestamp: int) -> dt.datetime:
    return dt.datetime.fromtimestamp(timestamp, tz=dt.UTC).replace(
        tzinfo=get_default_time_zone()
    )


class _DataConverter:
    def __init__(
        self,
        key: SensorKey,
        func: Callable[[CoordinatorData], SensorData],
    ) -> None:
        super().__init__()
        self.converter_key = key
        self.func = func

    def __call__(self, data: CoordinatorData) -> SensorData:
        return self.func(data)


def _data_converter[F: Callable[..., Any]](
    key: SensorKey,
) -> Callable[[F], _DataConverterT]:
    """Annotate and add a converter key to data converters.

    Returns:
        A decorator.
    """

    def wrapper(func: F) -> _DataConverterT:
        return cast(
            "_DataConverter",
            functools.wraps(func)(_DataConverter(key, func)),
        )

    return wrapper


@_data_converter(SensorKey.GALLONS_FOR_MOST_RECENT_HOUR)
def _sensor_data_for_most_recent_hour(data: CoordinatorData) -> SensorData:
    """Extract data for most recent hour.

    Returns:
        The extracted & converted records.
    """

    records = data["hourly"][-24:]
    record = records[-1]
    record_date = as_local(_from_timestamp(record["read_datetime"]))

    return {
        "state": _record_gallons(record),
        "attrs": {
            "start": record_date.isoformat(),
            "related": _serialize_records(records),
        },
    }


@_data_converter(SensorKey.GALLONS_FOR_MOST_RECENT_FULL_DAY_KEY)
def _sensor_data_for_most_recent_full_day(data: CoordinatorData) -> SensorData:
    """Extract data for first full day.

    Returns:
        The extracted & converted records.
    """

    records = _records_from_first_full_day(data)
    gallons = sum(_record_gallons(r) for r in records)

    return {
        "state": gallons,
        "attrs": {
            "related": _serialize_records(records),
        },
    }


@_data_converter(SensorKey.GALLONS_TODAY)
def _sensor_data_for_today(data: CoordinatorData) -> SensorData:
    """Extract data for today.

    Returns:
        The extracted & converted records.
    """

    records = _records_from_today(data)
    gallons = sum(_record_gallons(r) for r in records)

    return {
        "state": gallons,
        "attrs": {
            "related": _serialize_records(records),
        },
    }


def _records_from_today(data: CoordinatorData) -> list[UsageRecord]:
    """Extract records for today.

    Returns:
        The extracted records.
    """

    today_records = []
    start_of_day = start_of_local_day()

    for record in reversed(data["hourly"]):
        record_date = as_local(_from_timestamp(record["read_datetime"]))
        if record_date >= start_of_day:
            today_records.append(record)
        else:
            break

    return list(reversed(today_records))


def _records_from_first_full_day(data: CoordinatorData) -> list[UsageRecord]:
    """Extract records for first full day.

    Returns:
        The extracted records.
    """

    full_day_records = []
    last_full_day = None
    start_of_today = start_of_local_day()

    for record in reversed(data["hourly"]):
        record_date = as_local(_from_timestamp(record["read_datetime"]))
        start_of_day = start_of_local_day(record_date)

        if start_of_day >= start_of_today:
            continue

        if last_full_day and start_of_day < last_full_day:
            break

        if last_full_day and start_of_day == last_full_day:
            full_day_records.append(record)
        elif (
            not last_full_day
            and (record_date - start_of_day).total_seconds() // 3600 >= 23
        ):
            full_day_records.append(record)
            last_full_day = start_of_day

    return list(reversed(full_day_records))


def _record_gallons(record: UsageRecord) -> float | int:
    """Get record gallons guarded to ensure it's a number.

    Returns:
        The gallons or zero if it was not available.
    """

    result = record["gallons"]
    if result is None:
        result = 0
    return result


def _serialize_records(records: list[UsageRecord]) -> list[dict[str, Any]]:
    """Convert records for returning in attributes & service calls.

    Returns:
        The serialized records.
    """

    return [
        {
            "start": as_local(_from_timestamp(record["read_datetime"])).isoformat(),
            "gallons": _record_gallons(record),
        }
        for record in records
    ]
