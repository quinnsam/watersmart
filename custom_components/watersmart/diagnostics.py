"""Diagnostics support for WaterSmart."""

from typing import Any, cast

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant

from .coordinator import WaterSmartUpdateCoordinator

TO_REDACT = {
    CONF_USERNAME,
    CONF_PASSWORD,
}


async def async_get_config_entry_diagnostics(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: WaterSmartUpdateCoordinator = entry.runtime_data.coordinator

    return cast(
        "dict[str, Any]",
        async_redact_data(
            {
                "entry": entry.as_dict(),
                "data": coordinator.data,
            },
            TO_REDACT,
        ),
    )
