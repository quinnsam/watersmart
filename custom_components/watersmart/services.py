"""Support for the WaterSmart integration."""

from datetime import date, datetime
from functools import partial
from typing import Final, cast

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
    callback,
)
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import selector
from homeassistant.util import dt
import voluptuous as vol

from .const import DOMAIN
from .coordinator import (
    WaterSmartUpdateCoordinator,
    _from_timestamp,
    _serialize_records,
)
from .types import WaterSmartData

ATTR_CONFIG_ENTRY: Final = "config_entry"
ATTR_FROM_CACHE: Final = "cached"
ATTR_START: Final = "start"
ATTR_END: Final = "end"
HOURLY_HISTORY_SERVICE_NAME: Final = "get_hourly_history"

SERVICE_SCHEMA: Final = vol.Schema(
    {
        vol.Required(ATTR_CONFIG_ENTRY): selector.ConfigEntrySelector(
            {
                "integration": DOMAIN,
            }
        ),
        vol.Optional(ATTR_FROM_CACHE): bool,
        vol.Optional(ATTR_START): vol.Any(str, int),
        vol.Optional(ATTR_END): vol.Any(str, int),
    }
)


def __get_date(date_input: str | int | None) -> date | datetime | None:
    """Get date."""

    if not date_input:
        return None

    if isinstance(date_input, int) and (value := dt.utc_from_timestamp(date_input)):
        return cast(datetime, value)

    if isinstance(date_input, str) and (value := dt.parse_datetime(date_input)):
        return cast(datetime, value)

    raise ServiceValidationError(
        translation_domain=DOMAIN,
        translation_key="invalid_date",
        translation_placeholders={
            "date": date_input,
        },
    )


def __get_coordinator(
    hass: HomeAssistant, call: ServiceCall
) -> WaterSmartUpdateCoordinator:
    """Get the coordinator from the entry."""

    entry_id: str = call.data[ATTR_CONFIG_ENTRY]
    entry: ConfigEntry | None = hass.config_entries.async_get_entry(entry_id)

    if not entry:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="invalid_config_entry",
            translation_placeholders={
                "config_entry": entry_id,
            },
        )

    if entry.state != ConfigEntryState.LOADED:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="unloaded_config_entry",
            translation_placeholders={
                "config_entry": entry.title,
            },
        )

    runtime_data: WaterSmartData = hass.data[DOMAIN][entry_id]

    return runtime_data.coordinator


async def __get_hourly_history(
    call: ServiceCall,
    *,
    hass: HomeAssistant,
) -> ServiceResponse:
    coordinator = __get_coordinator(hass, call)

    start = __get_date(call.data.get(ATTR_START))
    end = __get_date(call.data.get(ATTR_END))

    if call.data.get(ATTR_FROM_CACHE) is False:
        await coordinator.async_refresh()

    records = []

    for record in coordinator.data["hourly"]:
        record_date = _from_timestamp(record["read_datetime"])

        if start and dt.as_local(record_date) < dt.as_local(start):
            continue

        if end and dt.as_local(record_date) > dt.as_local(end):
            continue

        records.append(record)

    return {"history": _serialize_records(records)}


@callback
def async_setup_services(hass: HomeAssistant) -> None:
    """Set up WaterSmart services."""

    hass.services.async_register(
        DOMAIN,
        HOURLY_HISTORY_SERVICE_NAME,
        partial(__get_hourly_history, hass=hass),
        schema=SERVICE_SCHEMA,
        supports_response=SupportsResponse.ONLY,
    )
