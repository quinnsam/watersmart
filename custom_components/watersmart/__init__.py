"""The WaterSmart integration."""

from __future__ import annotations

from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.typing import ConfigType

from .client import WaterSmartClient
from .const import DOMAIN
from .coordinator import WaterSmartUpdateCoordinator
from .services import async_setup_services
from .types import WaterSmartConfigEntry, WaterSmartData

PLATFORMS: list[Platform] = [Platform.SENSOR]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up WaterSmart services."""

    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: WaterSmartConfigEntry) -> bool:
    """Set up WaterSmart from a config entry."""

    hostname: str = entry.data[CONF_HOST]
    username: str = entry.data[CONF_USERNAME]
    password: str = entry.data[CONF_PASSWORD]

    session = async_get_clientsession(hass)
    watersmart = WaterSmartClient(hostname, username, password, session=session)

    coordinator = WaterSmartUpdateCoordinator(
        hass,
        watersmart,
        hostname,
        username,
        password,
    )

    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = WaterSmartData(
        coordinator=coordinator,
    )

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = entry.runtime_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: WaterSmartConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
