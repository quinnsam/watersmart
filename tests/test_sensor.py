"""Test sensor for simple integration."""

from homeassistant.const import UnitOfVolume
from homeassistant.core import HomeAssistant
from homeassistant.util.dt import utcnow

import datetime as dt
import pytest
from unittest.mock import patch, PropertyMock
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_fire_time_changed,
)
from typing import Generator

from custom_components.watersmart.client import AuthenticationError
from custom_components.watersmart.const import DOMAIN


class AdvacnedPropertyMock(PropertyMock):
    def __get__(self, obj, obj_type=None):
        return self(obj)

    def __set__(self, obj, val):
        self(obj, val)


@pytest.fixture
def mock_sensor_name() -> Generator[PropertyMock, None, None]:
    """Mock sensor names.

    This testing setup/library does not use `strings.json` and the entity description to translation key
    to get the entity name, so it's being patched here to just use the translaiton key. That way we at least
    get entity ids that are closer to what they will really be.
    """

    with patch(
        "homeassistant.components.sensor.SensorEntity.name",
        new_callable=AdvacnedPropertyMock,
    ) as mock_name:

        def name_from_entity_description(sensor):
            return sensor.entity_description.translation_key.replace(
                "_", " "
            ).capitalize()

        mock_name.side_effect = lambda self: (
            name_from_entity_description(self) if self else None
        )

        yield mock_name


async def test_most_recent_day_sensor(
    hass: HomeAssistant, mock_watersmart_client, mock_sensor_name
):
    """Test sensor."""
    hourly = mock_watersmart_client.async_get_hourly_data.return_value

    for gallons in range(25):
        last_time = dt.datetime.fromtimestamp(hourly[-1]["read_datetime"])
        next_time = int((last_time + dt.timedelta(hours=1)).timestamp())
        hourly.append(
            dict(
                hourly[-1],
                **{
                    "read_datetime": next_time,
                    "gallons": gallons,
                },
            )
        )

    mock_watersmart_client.async_get_hourly_data.return_value = hourly

    entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            "host": "test",
            "username": "test@home-assistant.io",
            "password": "Passw0rd",
        },
    )
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    recent_day_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_full_day"
    )

    assert recent_day_sensor_state
    assert recent_day_sensor_state.state == "1136"  # gallons changed to liters
    assert recent_day_sensor_state.attributes == {
        "attribution": "Data scraped from WaterSmart",
        "friendly_name": "WaterSmart (test) Gallons for most recent full day",
        "device_class": "water",
        "unit_of_measurement": UnitOfVolume.LITERS,
        "related": [
            {"start": "2024-06-20T00:00:00-07:00", "gallons": 1},
            {"start": "2024-06-20T01:00:00-07:00", "gallons": 2},
            {"start": "2024-06-20T02:00:00-07:00", "gallons": 3},
            {"start": "2024-06-20T03:00:00-07:00", "gallons": 4},
            {"start": "2024-06-20T04:00:00-07:00", "gallons": 5},
            {"start": "2024-06-20T05:00:00-07:00", "gallons": 6},
            {"start": "2024-06-20T06:00:00-07:00", "gallons": 7},
            {"start": "2024-06-20T07:00:00-07:00", "gallons": 8},
            {"start": "2024-06-20T08:00:00-07:00", "gallons": 9},
            {"start": "2024-06-20T09:00:00-07:00", "gallons": 10},
            {"start": "2024-06-20T10:00:00-07:00", "gallons": 11},
            {"start": "2024-06-20T11:00:00-07:00", "gallons": 12},
            {"start": "2024-06-20T12:00:00-07:00", "gallons": 13},
            {"start": "2024-06-20T13:00:00-07:00", "gallons": 14},
            {"start": "2024-06-20T14:00:00-07:00", "gallons": 15},
            {"start": "2024-06-20T15:00:00-07:00", "gallons": 16},
            {"start": "2024-06-20T16:00:00-07:00", "gallons": 17},
            {"start": "2024-06-20T17:00:00-07:00", "gallons": 18},
            {"start": "2024-06-20T18:00:00-07:00", "gallons": 19},
            {"start": "2024-06-20T19:00:00-07:00", "gallons": 20},
            {"start": "2024-06-20T20:00:00-07:00", "gallons": 21},
            {"start": "2024-06-20T21:00:00-07:00", "gallons": 22},
            {"start": "2024-06-20T22:00:00-07:00", "gallons": 23},
            {"start": "2024-06-20T23:00:00-07:00", "gallons": 24},
        ],
    }

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1


async def test_most_recent_hour_sensor(
    hass: HomeAssistant, mock_watersmart_client, mock_sensor_name
):
    """Test sensor."""
    mock_watersmart_client.async_get_hourly_data.return_value[-1]["gallons"] = 14.3

    entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            "host": "test",
            "username": "test@home-assistant.io",
            "password": "Passw0rd",
        },
    )
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    recent_hour_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_hour"
    )

    assert recent_hour_sensor_state
    assert recent_hour_sensor_state.state == "54.1"  # gallons changed to liters
    assert recent_hour_sensor_state.attributes == {
        "attribution": "Data scraped from WaterSmart",
        "friendly_name": "WaterSmart (test) Gallons for most recent hour",
        "device_class": "water",
        "unit_of_measurement": UnitOfVolume.LITERS,
        "start": "2024-06-19T22:00:00-07:00",
        "related": [
            {"gallons": 7.48, "start": "2024-06-19T19:00:00-07:00"},
            {"gallons": 0, "start": "2024-06-19T20:00:00-07:00"},
            {"gallons": 7.48, "start": "2024-06-19T21:00:00-07:00"},
            {"gallons": 14.3, "start": "2024-06-19T22:00:00-07:00"},
        ],
    }

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1


async def test_sensors_for_zero_gallons(
    hass: HomeAssistant, mock_watersmart_client, mock_sensor_name
):
    """Test sensor."""

    entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            "host": "test",
            "username": "test@home-assistant.io",
            "password": "Passw0rd",
        },
    )
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    recent_day_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_full_day"
    )

    assert recent_day_sensor_state
    assert recent_day_sensor_state.state == "0"

    recent_hour_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_hour"
    )

    assert recent_hour_sensor_state
    assert recent_hour_sensor_state.state == "0"

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1


async def test_sensor_update(
    hass: HomeAssistant, mock_watersmart_client, mock_sensor_name
):
    """Test sensor."""
    entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            "host": "test",
            "username": "test@home-assistant.io",
            "password": "Passw0rd",
        },
    )
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1

    async_fire_time_changed(hass, utcnow() + dt.timedelta(minutes=10))
    await hass.async_block_till_done()

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1

    async_fire_time_changed(hass, utcnow() + dt.timedelta(hours=1))
    await hass.async_block_till_done()

    assert mock_watersmart_client.async_get_hourly_data.call_count == 2


async def test_sensor_update_failure(
    hass: HomeAssistant, mock_watersmart_client, mock_sensor_name
):
    """Test sensor."""
    mock_watersmart_client.async_get_hourly_data.side_effect = AuthenticationError(
        "invalid credentials"
    )

    entry = MockConfigEntry(
        domain=DOMAIN,
        data={
            "host": "test",
            "username": "test@home-assistant.io",
            "password": "Passw0rd",
        },
    )
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    recent_day_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_full_day"
    )

    assert recent_day_sensor_state is None

    recent_hour_sensor_state = hass.states.get(
        "sensor.watersmart_test_gallons_for_most_recent_hour"
    )

    assert recent_hour_sensor_state is None

    assert mock_watersmart_client.async_get_hourly_data.call_count == 1
