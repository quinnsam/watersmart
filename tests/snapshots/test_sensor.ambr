# serializer version: 1
# name: test_most_recent_day_sensor
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data scraped from WaterSmart',
      'device_class': 'water',
      'friendly_name': 'WaterSmart (test) Gallons for most recent full day',
      'related': list([
        dict({
          'gallons': 1.0,
          'start': '2024-06-20T00:00:00-07:00',
        }),
        dict({
          'gallons': 2.0,
          'start': '2024-06-20T01:00:00-07:00',
        }),
        dict({
          'gallons': 3.0,
          'start': '2024-06-20T02:00:00-07:00',
        }),
        dict({
          'gallons': 4.0,
          'start': '2024-06-20T03:00:00-07:00',
        }),
        dict({
          'gallons': 5.0,
          'start': '2024-06-20T04:00:00-07:00',
        }),
        dict({
          'gallons': 6.0,
          'start': '2024-06-20T05:00:00-07:00',
        }),
        dict({
          'gallons': 7.0,
          'start': '2024-06-20T06:00:00-07:00',
        }),
        dict({
          'gallons': 8.0,
          'start': '2024-06-20T07:00:00-07:00',
        }),
        dict({
          'gallons': 9.0,
          'start': '2024-06-20T08:00:00-07:00',
        }),
        dict({
          'gallons': 10.0,
          'start': '2024-06-20T09:00:00-07:00',
        }),
        dict({
          'gallons': 11.0,
          'start': '2024-06-20T10:00:00-07:00',
        }),
        dict({
          'gallons': 12.0,
          'start': '2024-06-20T11:00:00-07:00',
        }),
        dict({
          'gallons': 13.0,
          'start': '2024-06-20T12:00:00-07:00',
        }),
        dict({
          'gallons': 14.0,
          'start': '2024-06-20T13:00:00-07:00',
        }),
        dict({
          'gallons': 15.0,
          'start': '2024-06-20T14:00:00-07:00',
        }),
        dict({
          'gallons': 16.0,
          'start': '2024-06-20T15:00:00-07:00',
        }),
        dict({
          'gallons': 17.0,
          'start': '2024-06-20T16:00:00-07:00',
        }),
        dict({
          'gallons': 18.0,
          'start': '2024-06-20T17:00:00-07:00',
        }),
        dict({
          'gallons': 19.0,
          'start': '2024-06-20T18:00:00-07:00',
        }),
        dict({
          'gallons': 20.0,
          'start': '2024-06-20T19:00:00-07:00',
        }),
        dict({
          'gallons': 21.0,
          'start': '2024-06-20T20:00:00-07:00',
        }),
        dict({
          'gallons': 22.0,
          'start': '2024-06-20T21:00:00-07:00',
        }),
        dict({
          'gallons': 23.0,
          'start': '2024-06-20T22:00:00-07:00',
        }),
        dict({
          'gallons': 24.0,
          'start': '2024-06-20T23:00:00-07:00',
        }),
      ]),
      'unit_of_measurement': <UnitOfVolume.LITERS: 'L'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.watersmart_test_gallons_for_most_recent_full_day',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1135.6',
  })
# ---
# name: test_most_recent_hour_sensor
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data scraped from WaterSmart',
      'device_class': 'water',
      'friendly_name': 'WaterSmart (test) Gallons for most recent hour',
      'related': list([
        dict({
          'gallons': 7.48,
          'start': '2024-06-19T19:00:00-07:00',
        }),
        dict({
          'gallons': 0,
          'start': '2024-06-19T20:00:00-07:00',
        }),
        dict({
          'gallons': 7.48,
          'start': '2024-06-19T21:00:00-07:00',
        }),
        dict({
          'gallons': 14.3,
          'start': '2024-06-19T22:00:00-07:00',
        }),
      ]),
      'start': '2024-06-19T22:00:00-07:00',
      'unit_of_measurement': <UnitOfVolume.LITERS: 'L'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.watersmart_test_gallons_for_most_recent_hour',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '54.1',
  })
# ---
# name: test_most_recent_hour_sensor_with_none_in_data
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data scraped from WaterSmart',
      'device_class': 'water',
      'friendly_name': 'WaterSmart (test) Gallons for most recent full day',
      'related': list([
        dict({
          'gallons': 1.0,
          'start': '2024-06-20T00:00:00-07:00',
        }),
        dict({
          'gallons': 2.0,
          'start': '2024-06-20T01:00:00-07:00',
        }),
        dict({
          'gallons': 3.0,
          'start': '2024-06-20T02:00:00-07:00',
        }),
        dict({
          'gallons': 4.0,
          'start': '2024-06-20T03:00:00-07:00',
        }),
        dict({
          'gallons': 5.0,
          'start': '2024-06-20T04:00:00-07:00',
        }),
        dict({
          'gallons': 6.0,
          'start': '2024-06-20T05:00:00-07:00',
        }),
        dict({
          'gallons': 7.0,
          'start': '2024-06-20T06:00:00-07:00',
        }),
        dict({
          'gallons': 8.0,
          'start': '2024-06-20T07:00:00-07:00',
        }),
        dict({
          'gallons': 9.0,
          'start': '2024-06-20T08:00:00-07:00',
        }),
        dict({
          'gallons': 10.0,
          'start': '2024-06-20T09:00:00-07:00',
        }),
        dict({
          'gallons': 11.0,
          'start': '2024-06-20T10:00:00-07:00',
        }),
        dict({
          'gallons': 12.0,
          'start': '2024-06-20T11:00:00-07:00',
        }),
        dict({
          'gallons': 13.0,
          'start': '2024-06-20T12:00:00-07:00',
        }),
        dict({
          'gallons': 14.0,
          'start': '2024-06-20T13:00:00-07:00',
        }),
        dict({
          'gallons': 15.0,
          'start': '2024-06-20T14:00:00-07:00',
        }),
        dict({
          'gallons': 16.0,
          'start': '2024-06-20T15:00:00-07:00',
        }),
        dict({
          'gallons': 17.0,
          'start': '2024-06-20T16:00:00-07:00',
        }),
        dict({
          'gallons': 18.0,
          'start': '2024-06-20T17:00:00-07:00',
        }),
        dict({
          'gallons': 19.0,
          'start': '2024-06-20T18:00:00-07:00',
        }),
        dict({
          'gallons': 20.0,
          'start': '2024-06-20T19:00:00-07:00',
        }),
        dict({
          'gallons': 0,
          'start': '2024-06-20T20:00:00-07:00',
        }),
        dict({
          'gallons': 22.0,
          'start': '2024-06-20T21:00:00-07:00',
        }),
        dict({
          'gallons': 23.0,
          'start': '2024-06-20T22:00:00-07:00',
        }),
        dict({
          'gallons': 24.0,
          'start': '2024-06-20T23:00:00-07:00',
        }),
      ]),
      'unit_of_measurement': <UnitOfVolume.LITERS: 'L'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.watersmart_test_gallons_for_most_recent_full_day',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1056.1',
  })
# ---
# name: test_sensors_for_zero_gallons
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data scraped from WaterSmart',
      'device_class': 'water',
      'friendly_name': 'WaterSmart (test) Gallons for most recent full day',
      'related': list([
      ]),
      'unit_of_measurement': <UnitOfVolume.LITERS: 'L'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.watersmart_test_gallons_for_most_recent_full_day',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '0',
  })
# ---
# name: test_sensors_for_zero_gallons.1
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data scraped from WaterSmart',
      'device_class': 'water',
      'friendly_name': 'WaterSmart (test) Gallons for most recent hour',
      'related': list([
        dict({
          'gallons': 7.48,
          'start': '2024-06-19T19:00:00-07:00',
        }),
        dict({
          'gallons': 0,
          'start': '2024-06-19T20:00:00-07:00',
        }),
        dict({
          'gallons': 7.48,
          'start': '2024-06-19T21:00:00-07:00',
        }),
        dict({
          'gallons': 0,
          'start': '2024-06-19T22:00:00-07:00',
        }),
      ]),
      'start': '2024-06-19T22:00:00-07:00',
      'unit_of_measurement': <UnitOfVolume.LITERS: 'L'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.watersmart_test_gallons_for_most_recent_hour',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '0',
  })
# ---
